xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java

```diff
@@ -0,0 +1,241 @@
+    private static final String MAPPINGS_DIR = File.separator + "mappings" + File.separator;
+    private static final String MAPPINGS_DIR_PATH = "%s" + MAPPINGS_DIR;
+    private final Aspsp aspsp;
+    private final ObjectMapper objectMapper;
+
+    public WiremockStubDifferenceDetectingInterceptor(Aspsp aspsp) {
+        this.aspsp = aspsp;
+        objectMapper = new ObjectMapper();
+    }
+
+    @Override
+    public Request.Builder apply(Request.Builder builder) {
+        return builder;
+    }
+
+    @Override
+    public <T> Response<T> postHandle(Request.Builder builder, Response<T> response) {
+        try {
+            WiremockFileResolver fileResolver = WiremockFileResolver.resolve(URI.create(builder.uri()).getPath(), builder.method(), builder.body());
+            String fileName = buildStubFilePath(aspsp.getName(), fileResolver.getFileName());
+            Map<String, Object> jsonFile = readStubFile(fileName);
+            List<String> changes = new ArrayList<>();
+            getStubRequestUrl(jsonFile)
+                .flatMap(url -> analyzeRequestUrl(fileResolver, builder, url))
+                .ifPresent(changes::add);
+            getStubRequestHeaders(jsonFile)
+                .flatMap(headers -> analyzeRequestHeaders(fileResolver, builder, headers))
+                .ifPresent(changes::add);
+            getRequestBody(jsonFile)
+                .flatMap(body -> analyzeRequestBody(fileResolver, builder, body))
+                .ifPresent(changes::add);
+            getResponseBody(jsonFile, aspsp.getName())
+                .flatMap(body -> analyzeResponseBody(fileResolver, response, body))
+                .ifPresent(changes::add);
+
+            Map<String, String> headersMap = response.getHeaders().getHeadersMap();
+
+            if (!changes.isEmpty()) {
+                String headerValue = String.join(",", changes);
+                headersMap.put(ResponseHeaders.X_ASPSP_CHANGES_DETECTED, headerValue);
+            }
+
+            return new Response<>(response.getStatusCode(), response.getBody(), ResponseHeaders.fromMap(headersMap));
+
+        } catch (Exception e) {
+            log.error("Can't find the difference with wiremock stub", e);
+        }
+
+        return response;
+    }
+
+    private <T> Optional<String> analyzeResponseBody(WiremockFileResolver resolver,
+                                                     Response<T> response,
+                                                     String responseBody) {
+        try {
+            String body = getResponseBody(response);
+            return analyzePayloadStructure(resolver, responseBody, body, PayloadType.RESPONSE);
+        } catch (JsonProcessingException e) {
+            log.error("Can't retrieve response body", e);
+        }
+        return Optional.empty();
+    }
+
+    @SuppressWarnings("unchecked")
+    private Optional<String> analyzePayloadStructure(WiremockFileResolver resolver, String payloadBody, String body, PayloadType type) {
```

Comment:
Have you considered using some libraries to do the diffing? E.g. xmlunit/jsonunit that wiremock itself is using. So that the comparison is performed beyond json keys. I'm not certain this method can even handle a json array.
Also manually finding the difference to analyse a warning will be quite labor intensive. Would be nice to have it already available in the log.

Reply:
Of course, I've looked into those libraries but as I understand they do not have the possibility to compare only the structure and I've found this solution on StackOverflow. Write differences into log

Reply:
I don't understand your answer. You're saying that a library makes a more thorough comparison and because of this you decided not to use it.

It looks like what you log as "differences" is just a body not a diff.

Reply:
For this functionality, we need to compare only the structure of a payload without the field values. But those libraries can't do that. That's why this functionality was written by ourselves.
Fixed logging object, thanks.

Reply:
I assume by "this functionality" you mean validation of wiremock mappings. Comparison of field values (types/sets of acceptable values) is necessary for that.

The logged object is still not a diff.

Reply:
3rd try.